### PR TITLE
Adds llamafile support for local LLM and embeddings

### PR DESCRIPTION
**Pull request by k8si**
# Description

Adds LLM and Embeddings clients for [llamafile](https://github.com/Mozilla-Ocho/llamafile). 

llamafile lets you run LLMs/embeddings locally from a single file on most computers without installing any dependencies.

To use the llamafile integrations, the user needs to:

1. Download a llamafile e.g. https://huggingface.co/jartine/TinyLlama-1.1B-Chat-v1.0-GGUF/resolve/main/TinyLlama-1.1B-Chat-v1.0.Q5_K_M.llamafile?download=true
2. Make the file executable.
3. Run the llamafile in 'server mode'. (All llamafiles come packaged with a lightweight server; by default, the server listens at http://localhost:8080.)

```bash
wget https://url/of/model.llamafile
chmod +x model.llamafile
./model.llamafile --server --nobrowser --embedding
```

Now, the user can invoke the LLM via the `llama-index` client:

```python
from llama_index.llms.llamafile import Llamafile

llm = Llamafile()
resp = llm.complete("Who is Octavia Butler?")
print(resp)
```

Or they can use the embeddings class:

```python
from llama_index.embeddings.llamafile import LlamafileEmbedding

embedder = LlamafileEmbedding()
text_embedding = embedder.get_text_embedding("Who is Octavia Butler?")
```

## Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-integrations/embeddings/llama-index-embeddings-llamafile/llama_index/embeddings/llamafile/base.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+from httpx import Timeout
+
+from llama_index.core.base.embeddings.base import BaseEmbedding, Embedding
+from llama_index.core.bridge.pydantic import Field
+from llama_index.core.callbacks.base import CallbackManager
+
+DEFAULT_REQUEST_TIMEOUT = 30.0
+
+
+class LlamafileEmbedding(BaseEmbedding):
```
Comment (logan-markewich) [inline llama-index-integrations/embeddings/llama-index-embeddings-llamafile/llama_index/embeddings/llamafile/base.py:13]: Do llama file models support batch embeddings? If so, we can implement `_get_text_embeddings()` and `_aget_text_embeddings()`

File: llama-index-integrations/llms/llama-index-llms-llamafile/llama_index/llms/llamafile/base.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+
+    @classmethod
+    def class_name(cls) -> str:
+        return "llamafile_llm"
+
+    @property
+    def metadata(self) -> LLMMetadata:
+        """LLM metadata."""
+        return LLMMetadata(
+            is_chat_model=True,  # llamafile has OpenAI-compatible chat API for all models
```
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-llamafile/llama_index/llms/llamafile/base.py:79]: If llama file is openai compatible, I wonder if it's possible to just extend the `OpenAILike` class here?

File: llama-index-integrations/embeddings/llama-index-embeddings-llamafile/poetry.lock
[the remark this answers is not in the document]
Comment (k8si) [reply llama-index-integrations/embeddings/llama-index-embeddings-llamafile/poetry.lock:1]: My bad. I added poetry.lock to the .gitignore for my packages so I don't accidentally do this again in the future.

File: llama-index-integrations/llms/llama-index-llms-llamafile/llama_index/llms/llamafile/base.py (hunk 2 above)
Comment (k8si) [reply llama-index-integrations/llms/llama-index-llms-llamafile/llama_index/llms/llamafile/base.py:79]: Was not aware of that class, I was just following the example of other integration classes like Ollama. 

The problem with extending `OpenAILike` here is that only the llamafile chat API `/v1/chat/completions` is OpenAI-compatible. The API calls involved in the `complete` method raise errors. 

I could override the `complete` methods but class will look like 50% the same as it is now and, when those methods get called, a lot of the internal plumbing (like the OpenAI client) just won't get used at all. 

Happy to go ahead and make this change if you want though

Comment (logan-markewich) [reply llama-index-integrations/llms/llama-index-llms-llamafile/llama_index/llms/llamafile/base.py:79]: Ah yea that's fair. No worries in that case

**Review by logan-markewich — APPROVED** [review]

**Pull request merged**